### ISSUE TITLE
Throw clean error if save() called in parallel on same doc instance

### ISSUE THREAD

**Issue by mtimofiiv** [Automattic/mongoose#4064]
I have seen a few issues closed already regarding the `#save()` method and swallowing of exceptions. I've got this happening at the moment with the latest version in npm (4.4.12). Here's some code:

``` js
const giftcard = new Giftcard({ // new instance of the model
  // setting some properties here
});

giftcard.save((err) => {
  if (err) throw new Error(err);
  // this never happens - the callback is never called
});
```

I can post a more detailed example if required.

Edit: Also, this happens with the promise interface as well, so `giftcard.save().then()` has the same result.

**Comment by mtimofiiv** [comment]
A little more insight into this bug.

 * Running `giftcard.save()` does not do the expected behaviour, but `giftcard.$__save()` does.
 * The output of `console.log(giftcard.save)` is `[Function: wrappedPointCut]` - this seems a little strange to me
 * When `giftcard.save()` is run, it does not execute inside the `save()` function - I tested this by adding a `console.log` inside the `save` method.

**Comment by vkarpov15** [comment]
@mtimofiiv are you setting `mongoose.Promise` anywhere? If I run the below script the script exits with an error as expected:

```javascript
'use strict';

var assert = require('assert');
var mongoose = require('mongoose');
var Schema = mongoose.Schema;

mongoose.connect('mongodb://localhost/gh4064');
mongoose.set('debug', true);

var schema = new mongoose.Schema({ name: String });
var Giftcard = mongoose.model('giftcard', schema);

const giftcard = new Giftcard({ // new instance of the model
  name: 'test'
});

giftcard.save((err) => {
  throw new Error('woops');
  // this never happens - the callback is never called
});
```

Output:

```
$ node gh-4064.js 
Mongoose: giftcards.insert({ name: 'test', _id: ObjectId("5712a91c21dd65a55d3fbfcc"), __v: 0 })   
/home/val/Workspace/10gen/troubleshoot-mongoose/node_modules/mongoose/node_modules/mongodb-core/lib/topologies/server.js:757
      catch(err) { process.nextTick(function() { throw err}); }
                                                 ^

Error: woops
    at /home/val/Workspace/10gen/troubleshoot-mongoose/gh-4064.js:18:9
    at model.<anonymous> (/home/val/Workspace/10gen/troubleshoot-mongoose/node_modules/mongoose/lib/document.js:1662:20)
    at next_ (/home/val/Workspace/10gen/troubleshoot-mongoose/node_modules/hooks-fixed/hooks.js:89:34)
    at fnWrapper (/home/val/Workspace/10gen/troubleshoot-mongoose/node_modules/hooks-fixed/hooks.js:186:18)
    at /home/val/Workspace/10gen/troubleshoot-mongoose/node_modules/mongoose/lib/model.js:269:13
    at /home/val/Workspace/10gen/troubleshoot-mongoose/node_modules/mongoose/lib/model.js:127:7
    at /home/val/Workspace/10gen/troubleshoot-mongoose/node_modules/mongoose/node_modules/mongodb/lib/collection.js:449:5
    at /home/val/Workspace/10gen/troubleshoot-mongoose/node_modules/mongoose/node_modules/mongodb/lib/collection.js:593:5
    at /home/val/Workspace/10gen/troubleshoot-mongoose/node_modules/mongoose/node_modules/mongodb/lib/bulk/unordered.js:457:9
    at resultHandler (/home/val/Workspace/10gen/troubleshoot-mongoose/node_modules/mongoose/node_modules/mongodb/lib/bulk/unordered.js:409:5)
    at /home/val/Workspace/10gen/troubleshoot-mongoose/node_modules/mongoose/node_modules/mongodb-core/lib/topologies/server.js:756:13
    at Callbacks.emit (/home/val/Workspace/10gen/troubleshoot-mongoose/node_modules/mongoose/node_modules/mongodb-core/lib/topologies/server.js:95:3)
    at null.messageHandler (/home/val/Workspace/10gen/troubleshoot-mongoose/node_modules/mongoose/node_modules/mongodb-core/lib/topologies/server.js:243:23)
    at Socket.<anonymous> (/home/val/Workspace/10gen/troubleshoot-mongoose/node_modules/mongoose/node_modules/mongodb-core/lib/connection/connection.js:262:22)
    at emitOne (events.js:77:13)
    at Socket.emit (events.js:169:7)
```

**Comment by mtimofiiv** [comment]
No, I am not overwriting the promise - I will endeavour to put together a more reproducable case for you.

**Comment by cwhenderson20** [comment]
@vkarpov15 I am experiencing this same issue on version 4.4.20. Attempting to call `save` on my model results in the callback never being called while running the same operation with `$__save` works just fine.

I am not overwriting promises and I do not have any pre-save (or otherwise) middleware for the model and using `mongoose.set("debug", true)` does not show anything strange. I have also verified that the document I'm attempting to save does pass validation.

I have additionally attached a connection error handler to the mongoose instance and no errors are reported.

Example:

```js
const appt = new Appointment();
// fill with data here

appt.save((err, doc) => {
    // callback never called
});

const appt2 = new Appointment();
// fill with data here

appt2.$__save({}, (err, doc) => {
    // callback is called
});
```

For now I will use the private method as I'm not sure of how to proceed otherwise, but if I can give any more information to help with this, please let me know.

**Comment by vkarpov15** [comment]
Can you provide a more concrete example, preferably with schemas? We've got a pretty wide range of tests for calling `.save()` and they're all passing...

**Comment by cwhenderson20** [comment]
Sure, here is the schema for the model that fails to save:

```js
const AppointmentSchema = new mongoose.Schema({
  patientId: {
    type: String,
    required: true
  },
  patientEmail: {
    type: String,
    trim: true,
    validate: [validator.isEmail, "Invalid email address"],
    required: true
  },
  patientName: {
    type: String,
    trim: true,
    required: true
  },
  doctor: {
    type: mongoose.Schema.ObjectId,
    ref: "User",
    required: true
  },
  date: {
    type: Date,
    required: true
  },
  sessionId: {
    type: String,
    required: true
  },
  updated: Date,
  canceled: Date,
  pushId: String
});
```

Save does not fail for this schema (at least not yet):
```js
const UserSchema = new mongoose.Schema({
  username: {
    type: String,
    lowercase: true,
    unique: true,
    required: true
  },
  email: {
    type: String,
    unique: true,
    lowercase: true,
    trim: true,
    required: true
  },
  fName: {
    type: String,
    trim: true,
    required: true
  },
  lName: {
    type: String,
    trim: true,
    required: true
  },
  hash: String,
  salt: String,
  role: {
    type: String,
    "enum": constants.USER_ROLES,
    "default": "concierge"
  }
});
```

As for the code that calls the save, it's simply a REST endpoint in express (some code is omitted for brevity and is unrelated):
```js
const mongoose = require("mongoose");
const Appointment = mongoose.model("Appointment");

app.post("/create", (req, res, next) => {
  const reqParams = ["patientId", "patientEmail", "patientName", "date", "doctor"];
  const appointment = new Appointment();

  reqParams.forEach((p) => appointment[p] = body[p]);
  appointment.save((err, appt) => {
    if (err) {
      return next(err);
    }

    res.json(appt);
  });
});
```

**Comment by mtimofiiv** [comment]
Similar situation here. Here's the schema:

```js
  Schemas.Giftcard = new Schema({
    project: { type: ObjectId, ref: 'Product', required: true },
    division: { type: ObjectId, ref: 'Division', required: true },
    slug: { type: String, required: true, index: true },
    name: { type: String, required: true },
    faceValue: { type: Number, required: true },
    tos: { type: ObjectId, ref: 'ContentElement' },
    image: { type: String, validate: [ validator.isURL, 'invalid URL' ] },
    endDate: { type: Date },
    oldID: { type: Number },
    referrerFields: [Schemas.Field],
    buyerFields: [Schemas.Field],
    contentElement: { type: ObjectId, ref: 'ContentElement' },
    redirectUri: { type: String, required: true, validate: [ validator.isURL, 'invalid URL' ] }
  });
```

And it's saved via an endpoint in an express api similar to @cwhenderson20:

```js
// DS is the object that holds all the models
const model = new DS.Giftcard(params);

model.save((err) => {
  if (err) return next(err);
});
```

**Comment by dodekeract** [comment]
Has anybody found out what triggers this? This particular bug seems to randomly appear in my codebase and I can't find out when and why.

**Comment by vkarpov15** [comment]
@cwhenderson20 below script works for me, can you modify and see if you can repro?

```javascript
'use strict';

var assert = require('assert');
var mongoose = require('mongoose');
var Schema = mongoose.Schema;

mongoose.connect('mongodb://localhost/gh4064');
mongoose.set('debug', true);

const AppointmentSchema = new mongoose.Schema({
  patientId: {
    type: String,
    required: true
  },
  patientEmail: {
    type: String,
    trim: true,
    validate: [function(v) { return true; }, "Invalid email address"],
    required: true
  },
  patientName: {
    type: String,
    trim: true,
    required: true
  },
  doctor: {
    type: mongoose.Schema.ObjectId,
    ref: "User",
    required: true
  },
  date: {
    type: Date,
    required: true
  },
  sessionId: {
    type: String,
    required: true
    required: true
  },
  updated: Date,
  canceled: Date,
  pushId: String
});

const Appointment = mongoose.model("Appointment", AppointmentSchema);

const body = { patientId: 'test', patientEmail: 'test', patientName: 'test', date: new Date(), doctor: new mongoose.Types.ObjectId() };
  const reqParams = ["patientId", "patientEmail", "patientName", "date", "doctor"];
  const appointment = new Appointment();

  reqParams.forEach((p) => appointment[p] = body[p]);
  appointment.sessionId = 'test';
  appointment.save((err, appt) => {
    assert.ifError(err);
    console.log('done', appt);
    process.exit(0);
  });
                                           
```

@dodekeract can you show me your schemas and your babel configs? I've never tested mongoose with babel and I wouldn't be surprised if they didn't play nicely together.

**Comment by dodekeract** [comment]
@vkarpov15 If I randomly get this issue again or manage to reproduce it, I'll get back to you. I've changed quite a lot in that particular codebase since it's a really new project and is still "finding its final structure", if you understand what I mean.

Regarding my Babel configs, I actually only use the CLI in a NPM script, and this is the command:

````
babel --plugins transform-async-to-generator --presets es2015,react source/ -d build/
````

Since Babel doesn't touch the mongoose files, I don't think it actually affects anything. I only mentioned it, because I'm usually using the Promise library from Babel (or is it in Node.js already?).

**Comment by vkarpov15** [comment]
@dodekeract the promise lib from babel is very different from the promise lib from node. How are you using it?

**Comment by dodekeract** [comment]
@vkarpov15 as I said "it doesn't matter if I remove/add my "custom" Promise library from Babel.". It makes literally no difference regarding this error.

**Comment by vkarpov15** [comment]
Ok thanks for clarifying. Can you show a more substantial code sample? I'm finding it hard to visualize what your code is structured like - are you using babel to compile code that imports mongoose?

**Comment by dodekeract** [comment]
Yes, you can find a non-bugged version of the code on my profile, impequid & impequid-dynamic-dns.

**Comment by dodekeract** [comment]
@vkarpov15 I can actually reproduce this issue in my project. In [this part](https://github.com/dodekeract/impequid/blob/master/source/server/database/users.js#L44-L67), if you change `$__save({}, (error, data) =>` to `save((error, data) =>` it stops working and doesn't fire any callbacks.

**Comment by mtimofiiv** [comment]
[My use case](https://github.com/phobosjs/phobos.js/blob/master/middleware/mutate.js#L66) (which I hotfixed to use `$__save`) does not have to do with promises also. The strange thing is some objects save and some do not.

My giftcard schema I shared earlier has the error but a different schema in the same app does not.

**Comment by vkarpov15** [comment]
Thanks @dodekeract , I'll try your code and see if I can repro. A test case would be very helpful :+1:

**Comment by dodekeract** [comment]
@vkarpov15 I know, I unfortunately have no idea why this is happening, so I can't really write a test.

**Comment by niftylettuce** [comment]
Also occurs in `4.4.20` and I'm sure others as well.  Using latest node stable 6.2.1.

**Comment by niftylettuce** [comment]
In my case, this error occurs when I set `req.user` equal to the `user` object returned from a `findOne` query.  Then later I call `req.user.someProp = 'foo';` and then `req.user.save(callback)` and similarly get this bug where the callback is not called, and if we `console.log` `req.user.save` it outputs the wrappedPointCut function.

**Comment by niftylettuce** [comment]
I implemented a hotfix sitewide for my app, in my case here's what I did:

```bash
npm i replace
node_modules/.bin/replace "\.save\(" ".\$__save({}, " . -r --exclude node_modules
```

This will replace all instances of `.save` with `.$__save` and a blank options object.

**Comment by vkarpov15** [comment]
@niftylettuce can you show me your schema?

**Comment by niftylettuce** [comment]
@vkarpov15 its very simple and it has pre validate and pre save hooks, I can't share the full schema for privacy, but here's a snippet:

```js
var passportLocalMongoose = require('passport-local-mongoose')
var jsonSelect = require('mongoose-json-select')

// the below is a very simplified version of the full schema which is extensive
var User = new Schema({
  group: String,
  referral_code_url: String,
  full_name: String,
  full_email: String,
});

User.pre('save', function(next) {

  console.log('saving user avatar');

  var user = this

  // set default avatar_url to gravatar
  if ( (!_.isString(user.avatar_url) || _.isEmpty(user.avatar_url)) && validator.isEmail(user.email)) {
    var gravatar = crypto.createHash('md5').update(this.email).digest('hex')
    user.avatar_url = 'https://secure.gravatar.com/avatar/' + gravatar
  }

  next()

});

// ...
// more pre save and pre validate hooks...
// ...

User.plugin(passportLocalMongoose, {
  usernameField: 'email',
  usernameLowerCase: true,
  userExistsError: 'User already exists with email %s',
  populateFields: 'some_tier'
})

User.plugin(jsonSelect, '-_group -salt -hash')
```

Perhaps passport local mongoose (in particular the `populateFields` option) or jsonSelect causes the issue?

**Comment by niftylettuce** [comment]
okay so apparently if you use `.$__save` the hooks don't get called, aghhh!!

**Comment by niftylettuce** [comment]
This also affects version 4.3.x

**Comment by niftylettuce** [comment]
Also affects version 4.2.x

**Comment by niftylettuce** [comment]
Seems like Node 6.x affects this somehow?

**Comment by niftylettuce** [comment]
@vkarpov15 I will PayPal $150 if you can fix this today.

**Comment by vkarpov15** [comment]
@niftylettuce wish I could help but the below script works fine for me (node 6.2.2, mongoose 4.4.20):

```javascript
'use strict';

var assert = require('assert');
var mongoose = require('mongoose');
var Schema = mongoose.Schema;

mongoose.connect('mongodb://localhost/gh4064');
mongoose.set('debug', true);

var _ = require('lodash');
var passportLocalMongoose = require('passport-local-mongoose')
var jsonSelect = require('mongoose-json-select')
var validator = require('validator');
var crypto = require('crypto');

// the below is a very simplified version of the full schema which is extensive
var User = new Schema({
  group: String,
  referral_code_url: String,
  full_name: String,
  full_email: String,
});

User.pre('save', function(next) {

  console.log('saving user avatar');

  var user = this

  // set default avatar_url to gravatar
  if ( (!_.isString(user.avatar_url) || _.isEmpty(user.avatar_url)) && validator.isEmail(user.email)) {
    var gravatar = crypto.createHash('md5').update(this.email).digest('hex')
    user.avatar_url = 'https://secure.gravatar.com/avatar/' + gravatar
  }

  next()

});

// ...
// more pre save and pre validate hooks...
// ...

User.plugin(passportLocalMongoose, {
  usernameField: 'email',
  usernameLowerCase: true,
  userExistsError: 'User already exists with email %s',
  populateFields: 'some_tier'
})

User.plugin(jsonSelect, '-_group -salt -hash')

var U = mongoose.model('User', User);

U.create({ email: 'test@bacon.co' }, function(error, user) {
  assert.ifError(error);

  U.findOne({ _id: user._id }).exec(function(error, user) {
    user.name = 'test';
    user.save(function(error) {
      assert.ifError(error);
      console.log('done');
      process.exit(0);
    });
  });
});
```

Unless I can repro, it's really difficult for me to debug. Maybe you can tweak the above script to see if you can trigger the error?

Also, FYI, you're right, calling `$__save()` will skip all hooks, that's where mongoose puts the original save function when wrapping things in hooks.

**Comment by vkarpov15** [comment]
Hmm @dodekeract so I tried running your test suite with changing `$__save()` to `save()` in https://github.com/phobosjs/phobos.js/blob/master/middleware/mutate.js#L66 and it succeeds... Does it fail for you if you make this change?

**Comment by dodekeract** [comment]
@vkarpov15 sorry, I don't really understand what I'm supposed to try.

**Comment by niftylettuce** [comment]
Perhaps this is related to a specific MongoDB version?

**Comment by niftylettuce** [comment]
@vkarpov15 and I talked over Google Chat, and we discovered that due to the `validator` library accepting two arguments for most functions, e.g. my case with `isEmail(str, options)` and @mtimofiiv case with `validator.isURL(str, options)`, it thinks that the `options` object is the asynchronous callback, which it isn't a function, so nothing happens and it hangs.

**Comment by mtimofiiv** [comment]
@niftylettuce wow that does make a lot of sense. I will test this to confirm!

**Comment by dodekeract** [comment]
Quite sure that this will solve the issues I had. At one point I was already wondering why `validate: v => isIP(v)` would fix a similar issue .

**Comment by vkarpov15** [comment]
Yeah it all comes back to the somewhat annoying implicit assumption that validators that take 2 params are async. As a workaround in 4.5.3, we check the return value of the validator and, if it's a boolean, we just use that and prevent the callback from getting fired. Somewhat risky but a lot of people have gotten bit by this and it's a really hard bug to track down as @niftylettuce and I learned...

**Comment by niftylettuce** [comment]
@mtimofiiv @cwhenderson20 @dodekeract I have just donated over PayPal to @vkarpov15 (val@karpov.io) for his time helping me debug this issue.  If you have been able to get passed this through our findings please send him some dough! 💸 He deserves it!  He spent a ton of time with me, even while he was on his mobile phone he was helping!

**Comment by HenrikGr** [comment]
I'm not using validator but getting this myself.
Can anybode huelp me, has posted more details on stack overflow.

http://stackoverflow.com/questions/40315295/cant-create-new-entity-in-mongodb-experimental-ibm-bluemix

**Comment by newtonkr** [comment]
This code works for inserting first record thenAfter it gives error inserting in database ie 400

router.post('/child_register', function(req,res,next)
{
   var childInfo = req.body;
   if(!childInfo.phone || !childInfo.password|| !childInfo.parent_phone)
   {
      res.render('show_message', {message: "Sorry, you provided worng info", type: "error"});
   }
   else
   {
      Child.findOne(
      {phone:req.body.phone},function(err, response)
      {
             if(response==null)
             {

                          var newChild = new Child({name:childInfo.name,phone:childInfo.phone,parent_phone:"null",password:childInfo.password,wallet:0});
                          var newPairNotification = new Pair({childphone:childInfo.phone,parentphone:childInfo.parent_phone});
                          Parent.findOne(
                          {phone:childInfo.parent_phone},function(err, response1)
                          {
                          if(response1==null)
                          {

                                newChild.$__save({},function(err,Child)
                                {

                                  if(err)
                                  {

                                      res.json(400);
                                  }
                                  else
                                  {

                                      newPairNotification.save(function(err1)
                                      {
                                      if(err1)
                                      {

                                      res.json(600);
                                      }
                                      else
                                      {

                                      res.json(201);//'show_message', {message: "New person added", type: "success", person: personInfo});
                                      }
                                      });

                                  }
                                });

                          }
                          else
                          {
                                //,parent_phone:childInfo.parent_phone
                                newChild.save(function(err,Child)
                                {

                                    if(err)
                                    {
                                        res.json(401);
                                    }
                                    else
                                    {
                                        newPairNotification.save(function(err1,Pair)
                                        {
                                          if(err1)
                                          {
                                              res.json(600);
                                          }
                                          else
                                          {
                                              res.json(201);//'show_message', {message: "New person added", type: "success", person: personInfo});
                                          }
                                        });
                                    }
                                });

                          }

                          });


            }
            else
            {
               res.json(44);
            }

      });

      }

});

**Comment by varunjayaraman** [comment]
@newtonkr please open a separate issue or explain how your issue is related to the current issue you're posting in. Also we need more info: what is the stack trace and if you could please a) format your code so its easy to read and b) if you could avoid posting a huge blob of code if you know around where the problem exists, that would be greatly appreciated.

**Comment by newtonkr** [comment]
my error resolved actually i was making schema and then connecting it to mongodb that was causing problem in Model.save()

**Comment by ketimaBU** [comment]
I am getting the same issue without using two params

**Comment by scalpo** [comment]
this solved my problem... connect using mongo client...
http://mongoosejs.com/docs/connections.html#use-mongo-client

**Comment by mayeaux** [comment]
I'm getting this too randomly?

**Comment by vkarpov15** [comment]
@mayeaux please provide some code samples and verify that you're not experiencing any connection outages. See http://thecodebarbarian.com/mongoose-4.12-improved-connection-events.html for some more info.

**Comment by RaphiStein** [comment]
I was having this issue also. My code looked like 

`userInstance.save((err, doc) => {...})`

I fixed it by changing to 

```
const promise = userInstance.save();
Promise.all([promise]).then( function(){ ... })
```

works now!

**Comment by chaddjohnson** [comment]
@RaphiStein 

Why are you doing 

```javascript
const promise = userInstance.save();
Promise.all([promise]).then( function(){ ... })
```

as opposed to

```javascript
userInstance.save().then( function(){ ... })
```

Can you explain what advantage using `Promise.all([promise])` offers? I am getting a `wrappedPointCut` error as well.

**Comment by vkarpov15** [comment]
@chaddjohnson there shouldn't be any difference between the two. Can you clarify what error message you're getting?

**Comment by chaddjohnson** [comment]
@vkarpov15 Thanks for the wonderful library. It's fantastic.

So, very occasionally I get errors like this:

```
Error
    at model.wrappedPointCut [as save] (/home/project/build/processes/1517989482/node_modules/mongoose/lib/services/model/applyHooks.js:131:29)
    at generator.generate.then.generatedImageUrls (/home/project/build/processes/1517989482/node_modules/@chaddjohnson/services/lib/models/product.js:119:25)
    at <anonymous>
```

```
Error
    at model.wrappedPointCut [as save] (/home/project/build/processes/1518143934/node_modules/mongoose/lib/services/model/applyHooks.js:131:29)
    at handleOrderProcessing.then.catch.transactionError (/home/project/build/processes/1518143934/node_modules/@chaddjohnson/services/lib/paymentProcessors/authorizeNet.js:203:34)
    at <anonymous>
    at process._tickDomainCallback (internal/process/next_tick.js:228:7)
```

When this happens, it happens when I call `.save()` on a model instance. I have no solid leads as to the cause. The only things I can think of are:

1) I am using `usePushEach` for every model.
2) I sometimes bypass Mongoose `.save()` to prevent triggering post save hooks, like so:

```javascript
return new Promise((resolve, reject) => {
    Order.update({_id: order._id, __v: order.__v}, {$set: {shopifyOrderData: data.shopifyOrderData}}, (error) => {
        if (error) {
            return reject(error);
        }

        resolve();
    });
})
```

**Comment by chaddjohnson** [comment]
@vkarpov15 Here you go:



An instance method within a class (note the comment `THE ERROR IS TRIGGERED HERE`):

```javascript
processOrder(orderData) {
    return this.handleOrderProcessing(orderData)
        .then(transactionData => {
            // Processing succeeded, so reset the payment processor failure count.
            this.config.failureCount = 0;

            // Update the shop, and return the transaction data.
            return this.shop.save()
                .then(() => transactionData)
                .catch(error => {
                    loggerService.error(`Shop ${this.shop.domain} was charged via ${this.getName()} for order ${orderData.id}, but resetting failure count failed`, error);
                    return transactionData;
                });
        })
        .catch(transactionError => {
            // Processing failed, so increment the payment processor failure count.
            this.config.failureCount++;

            // Disable the payment processor if it has failed too many times.
            if (this.config.failureCount >= 3) {
                this.config.disabled = true;

                loggerService.info(`${this.getName()} payment method for shop ${this.shop.domain} disabled as failure count has reached maximum allowed`);
            }

            // Update the shop, and re-throw the error.
            return this.shop.save()  // THE ERROR IS TRIGGERED HERE ("wrappedPointCut [as save]")
                .catch(error => {
                    loggerService.error(`Shop ${this.shop.domain} was NOT charged via ${this.getName()} for order ${orderData.id}, and incrementing failure count failed`, error);
                    throw transactionError;
                })
                .then(() => Promise.reject(transactionError));
        });
}
```

Within my `Shop` model (which calls the method):

```javascript
schema.methods.processPayment = function(paymentProcessorOrderData) {
    const paymentProcessors = this.getPaymentProcessors();

    // ...

    /*
     * Run processors in order, one by one, until one succeeds or all
     * fail. This will only run the next payment processor if the current
     * one fails.
     */
    return paymentProcessors.reduce((current, next) => {
        // Run the current payment processor.
        return current.catch(error => {
            if (error && currentPaymentProcessor) {
                // Log that payment processing failed using the payment processor.
                loggerService.warn(`Batch payment processing of \$${paymentProcessorOrderData.orderTotal} for shop ${this.domain} failed using ${currentPaymentProcessor.getName()}`);
            }

            currentPaymentProcessor = next;

            // Try the next payment processor.
            return next.processOrder(paymentProcessorOrderData);
        });
    }, Promise.reject())
        .catch(error => {
            if (currentPaymentProcessor) {
                // Log that payment processing failed using the last payment processor.
                loggerService.warn(`Batch payment processing of \$${paymentProcessorOrderData.orderTotal} for shop ${this.domain} failed using ${currentPaymentProcessor.getName()}`);
            }

            // ...

            throw error;
        });
}
```

**Comment by rkm660** [comment]
For me, the easiest workaround was to use "insertMany" with an array of one doc. Logging my schema object showed no save function for some reason.

**Comment by uptownhr** [comment]
all, i know this issue is marked closed but I'm seeing this issue right now in my tests (using ava). Ava runs tests concurrently and I believe it is related to the same model instance saving concurrently. Anyone else have insights into this?

**Comment by vkarpov15** [comment]
@uptownhr do you have one `doc` that you're calling `save()` on in multiple tests? Why not just create a new doc for every test?

**Comment by uptownhr** [comment]
@vkarpov15 you're correct. I did do that to work around the issue. 

Just curious though, is this the expected behavior?

**Comment by vkarpov15** [comment]
What behavior specifically are you referring to?

In general, saving the same document reference multiple times in parallel leads to unpredictable results, and you shouldn't need to do that...

**Comment by vkarpov15** [comment]
Because `save()` has a bunch of async steps: validation, middleware, etc, so if you `save()` the same doc twice in parallel you might cases where two of these async tasks conflict with each other, like validating and then invalidating the same path. We should consider throwing an error in this case.

**Comment by melunar** [comment]
我这边也遇到了save方法不生效, promise状态一直pending的问题, 不过找了很久发现问题是因为: 我项目的mongo连接方法使用mongo原生connect方法写的, 业务api使用mongoose操作的model, 我把连接方法改成mongoose.connect就好了, 有点荒唐, 不知道对大家有没有帮助.

**Comment by mourice-oduor** [comment]
Hello guys I'm new to Nodejs, while using mongoose, I've encountered this callback function error which fails to save models on my DB: 
var mongoose = require("mongoose")
mongoose.connect("mongodb://localhost/cat_app", { useNewUrlParser: true, useUnifiedTopology: true});

var catSchema = new mongoose.Schema({
    name: String,
    age: Number,
    temparament: String
})

var Cat = mongoose.model("Cat", catSchema)

//adding a new cat to the DB
var george = new Cat({
    name: "George",
    age: 10,
    temparament: "Grouchy"
})

// george.$__save({}, (err, cat) =>  {
// george.save((err, cat) => {})
george.save(function (err, cat) {
    if(err){
        console.log("SOMETHING WENT WRONG")
    } else {
        console.log("WE JUST SAVED A CAT TO THE DB:")
        console.log(cat)
    }
        
})

**Comment by vkarpov15** [comment]
Did you accidentally uncomment `george.save((err, cat) => {})`? Calling `save()` twice in a row would trigger this error.

**Comment by mourice-oduor** [comment]
Yes I did so

**Comment by mourice-oduor** [comment]
I came to found the error was that I had enabled security,  after disabling
it then it worked out well... Thanks guys
